Bottom: ff1d3f1895b0f5b7073a5c31cc0d287c1e0c051a
Top:    b34661bd897e6a163b13168475e826b7edaf7ede
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-06-26 09:17:30 -0500

Use docker named volumes

Instead of a data container, use docker named volume

---
diff --git a/ansible-playbook/library/docker_datavolume b/ansible-playbook/library/docker_datavolume
deleted file mode 100644
index f79916f1..00000000
--- a/ansible-playbook/library/docker_datavolume
+++ /dev/null
@@ -1,54 +0,0 @@
-#!/bin/bash
-
-# Ansible transports arguments to modules in a file. The
-# path to the arguments file is in $1, and the file
-# contains the module's arguments like this:
-#
-#       image_name=novatechweb/ldap image_tag=current data_volume_container_name=ldap_DV
-#
-# Arguments:
-#    image_name
-#    image_tag
-#    data_volume_container_name
-
-eval $(sed -e "s/\s?\([^=]+\)\s?=\s?\(\x22\([^\x22]+\)\x22|\x27\([^\x27]+\)\x27|\(\S+\)\)\s?/\1='\2'/p" $1)
-
-if [[ -z "${image_name}" ]]; then
-    printf "{\"failed\": true, \"changed\": false, \"msg\": \"Module needs image_name= argument\"}"
-    exit 1
-fi
-
-if [[ -z "${image_tag}" ]]; then
-    printf "{\"failed\": true, \"changed\": false, \"msg\": \"Module needs image_tag= argument\"}"
-    exit 1
-fi
-
-if [[ -z "${data_volume_container_name}" ]]; then
-    printf "{\"failed\": true, \"changed\": false, \"msg\": \"Module needs data_volume_container_name= argument\"}"
-    exit 1
-fi
-
-# *****************************************************************************
-# Check if data_volume_container_name exists
-
-docker inspect ${data_volume_container_name} &> /dev/null
-result="${?}"
-
-if [[ ${result} == "0" ]]; then
-    printf "{\"failed\": false, \"changed\": false, \"msg\": \"DataVolume container name already exists: ${data_volume_container_name}\"}"
-    exit 0
-fi
-
-# *****************************************************************************
-# create the container for the Data Volume
-
-error_msg=$(docker 2>&1 run --name ${data_volume_container_name} --entrypoint /bin/true ${image_name}:${image_tag})
-result="${?}"
-
-if [[ ${result} != "0" ]]; then
-    printf "{\"failed\": true, \"changed\": false, \"msg\": \"${error_msg}\"}"
-    exit 1
-fi
-
-printf "{\"failed\": false, \"changed\": true, \"msg\": \"Created container ${data_volume_container_name} for the Data Volume\"}"
-exit 0
diff --git a/docker/docker-mantisbt/mantisbt.sh b/docker/docker-mantisbt/mantisbt.sh
index 7c584d4d..44dc2c31 100755
--- a/docker/docker-mantisbt/mantisbt.sh
+++ b/docker/docker-mantisbt/mantisbt.sh
@@ -56,8 +56,8 @@ done
 # make certian the containers exist
 docker inspect ${MANTISBT_CONTAINER_NAME} > /dev/null
 docker inspect ${MANTISBT_DB_CONTAINER_NAME} > /dev/null
-docker inspect ${MANTISBT_DV_NAME} > /dev/null
-docker inspect ${MANTISBT_DB_DV_NAME} > /dev/null
+docker volume inspect ${MANTISBT_DV_NAME} > /dev/null
+docker volume inspect ${MANTISBT_DB_DV_NAME} > /dev/null
 
 get_db_user_and_password() {
     db_user="$(docker 2>&1 exec ${MANTISBT_CONTAINER_NAME} grep -e '^$g_db_username' config_inc.php|sed 's|^.* = \"\(.*\)\";$|\1|' || true)"
diff --git a/docker/docker-mediawiki/mediawiki.sh b/docker/docker-mediawiki/mediawiki.sh
index d52bd317..be0f4f07 100755
--- a/docker/docker-mediawiki/mediawiki.sh
+++ b/docker/docker-mediawiki/mediawiki.sh
@@ -53,8 +53,8 @@ done
 # make certian the containers exist
 docker inspect ${WIKI_CONTAINER_NAME} > /dev/null
 docker inspect ${WIKI_DB_CONTAINER_NAME} > /dev/null
-docker inspect ${WIKI_DV_NAME} > /dev/null
-docker inspect ${WIKI_DB_DV_NAME} > /dev/null
+docker volume inspect ${WIKI_DV_NAME} > /dev/null
+docker volume inspect ${WIKI_DB_DV_NAME} > /dev/null
 
 get_db_user_and_password() {
     db_user="$(docker 2>&1 exec ${WIKI_CONTAINER_NAME} grep -e '^$wgDBuser' LocalSettings.php|sed 's|^.* = \"\(.*\)\";$|\1|' || true)"
diff --git a/roles/docker-exim4/defaults/main.yml b/roles/docker-exim4/defaults/main.yml
index 8a1f2ef2..087ad336 100644
--- a/roles/docker-exim4/defaults/main.yml
+++ b/roles/docker-exim4/defaults/main.yml
@@ -23,6 +23,9 @@ exim4_image_name: '{{ docker_registry_username }}/exim4'
 # the name of the container being started
 exim4_container_name: '{{ docker_name_prefix }}exim4'
 
+# the name of the config-volume used by the container
+exim4_cv_name: '{{ docker_name_prefix }}exim4_CV'
+
 # the name of the data-volume used by the container
 exim4_dv_name: '{{ docker_name_prefix }}exim4_DV'
 
diff --git a/roles/docker-exim4/tasks/main.yml b/roles/docker-exim4/tasks/main.yml
index d670d37d..3cebb60d 100644
--- a/roles/docker-exim4/tasks/main.yml
+++ b/roles/docker-exim4/tasks/main.yml
@@ -92,10 +92,12 @@
 # Create the data volumes
 
 - name: data-volume container
-  docker_datavolume:
-    image_name: '{{ exim4_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ exim4_dv_name }}'
+  docker_volume:
+    name: '{{ exim4_cv_name }}'
+
+- name: data-volume container
+  docker_volume:
+    name: '{{ exim4_dv_name }}'
 
 # *****************************************************************************
 # Start the data container running
@@ -109,6 +111,7 @@
     restart_policy: '{{ docker_restart_policy }}'
     state: started
     volumes:
+      - '{{ exim4_cv_name }}:/etc/exim4:z'
       - '{{ exim4_dv_name }}:/var/spool/exim4:z'
       - '{{ openssl_dv_name }}:/etc/ssl/private:z'
 
diff --git a/roles/docker-gitlab/tasks/main.yml b/roles/docker-gitlab/tasks/main.yml
index 3957c3a2..387e7adf 100644
--- a/roles/docker-gitlab/tasks/main.yml
+++ b/roles/docker-gitlab/tasks/main.yml
@@ -97,16 +97,12 @@
 # Create the data volumes
 
 - name: data-volume container (gitlab)
-  docker_datavolume:
-    image_name: '{{ gitlab_dv_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ gitlab_dv_name }}'
+  docker_volume:
+    name: '{{ gitlab_dv_name }}'
 
 - name: data-volume container (postgres)
-  docker_datavolume:
-    image_name: '{{ gitlab_db_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ gitlab_db_dv_name }}'
+  docker_volume:
+    name: '{{ gitlab_db_dv_name }}'
 
 - name: stop prev container (postgres)
   docker_container:
diff --git a/roles/docker-mantisbt/tasks/main.yml b/roles/docker-mantisbt/tasks/main.yml
index 45c1a895..a8ff7fa4 100644
--- a/roles/docker-mantisbt/tasks/main.yml
+++ b/roles/docker-mantisbt/tasks/main.yml
@@ -113,16 +113,12 @@
 # Create the data volumes
 
 - name: data-volume container (mantisbt)
-  docker_datavolume:
-    image_name: '{{ mantisbt_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ mantisbt_dv_name }}'
+  docker_volume:
+    name: '{{ mantisbt_dv_name }}'
 
 - name: data-volume container (mysql)
-  docker_datavolume:
-    image_name: '{{ mantisbt_db_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ mantisbt_db_dv_name }}'
+  docker_volume:
+    name: '{{ mantisbt_db_dv_name }}'
 
 - name: stop prev container (mysql)
   docker_container:
@@ -162,6 +158,7 @@
     detach: true
     restart_policy: '{{ docker_restart_policy }}'
     hostname: '{{ container_addr_map.mantisbt.hostname }}'
+    hostname: '{{ mantisbt_hostname }}'
     ports: '{{ container_port_map.mantisbt.port_args }}'
     link: 
       -'{{ openldap_container_name }}:mantisbt_ldap'
@@ -171,10 +168,10 @@
       - '{{ openssl_dv_name }}'
       - '{{ mantisbt_dv_name }}'
     env: 
-      - MANTISBT_HOSTNAME={{ container_addr_map.mantisbt.hostname }}
-      - MANTISBT_MAIL_USER=mantis
-      - MANTISBT_MAIL_PASSWORD={{ mantisbt_email_password | quote }}
-      - MANTISBT_LDAP_PASSWORD={{ ldap_proxyagent_password | quote }}
+      MANTISBT_HOSTNAME: '{{ container_addr_map.mantisbt.hostname }}'
+      MANTISBT_MAIL_USER: 'mantis'
+      MANTISBT_MAIL_PASSWORD: '{{ mantisbt_email_password | quote }}'
+      MANTISBT_LDAP_PASSWORD: '{{ openldap_proxyagent_password | quote }}'
     image: '{{ mantisbt_image_name }}:{{ docker_image_tag }}'
 
 # *****************************************************************************
diff --git a/roles/docker-mediawiki/defaults/main.yml b/roles/docker-mediawiki/defaults/main.yml
index a95335ca..2a86d85e 100644
--- a/roles/docker-mediawiki/defaults/main.yml
+++ b/roles/docker-mediawiki/defaults/main.yml
@@ -24,7 +24,7 @@ wiki_image_repo: https://github.com/novatechweb/docker-mediawiki.git
 
 # the name of the image being duilt and used for the container
 wiki_image_name: '{{ docker_registry_username }}/wiki'
-wiki_db_image_name: '{{ repository_images.mysql.image }}'
+wiki_db_image_name: 'mysql:5'
 
 # the name of the container being started
 wiki_container_name: '{{ docker_name_prefix }}wiki'
@@ -45,3 +45,7 @@ wiki_docker_restore_dir: '{{ bacula_dest }}{{ wiki_docker_backup_dir }}'
 # files restored from tape
 wiki_static_backup_file: '/mediawiki.tar'
 wiki_database_backup_file: '/wikidb.sql'
+
+
+# the name of the sql database used by mediawiki
+wiki_database_name: wikidb
diff --git a/roles/docker-mediawiki/tasks/main.yml b/roles/docker-mediawiki/tasks/main.yml
index cb19329d..0a9243fb 100644
--- a/roles/docker-mediawiki/tasks/main.yml
+++ b/roles/docker-mediawiki/tasks/main.yml
@@ -87,44 +87,46 @@
 # Create the data volumes
 
 - name: data-volume container (mediwiki)
-  docker_datavolume:
-    image_name: '{{ wiki_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ wiki_dv_name }}'
+  docker_volume:
+    name: '{{ wiki_dv_name }}'
 
 - name: data-volume container (mysql)
-  docker_datavolume:
-    image_name: '{{ wiki_db_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ wiki_db_dv_name }}'
-
-- name: stop prev container (mysql)
-  docker_container:
-    image: '{{ wiki_db_image_name }}'
-    name: '{{ wiki_db_container_name }}'
-    state: stopped
-
-- name: initial populate (mysql)
-  docker_db_init:
-    image_name: '{{ wiki_db_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    container_name: '{{ wiki_db_dv_name }}_init_db'
-    data_volume_container_name: '{{ wiki_db_dv_name }}'
-    database_name: wikidb
-    database_user: '{{ wiki_db_user }}'
-    database_password: '{{ wiki_db_password | quote }}'
-    database_root_password: '{{ wiki_db_root_password | quote }}'
+  docker_volume:
+    name: '{{ wiki_db_dv_name }}'
 
 # *****************************************************************************
-# Start the data container running
+# Start the database
 
 - name: start container (mysql)
   docker_container:
     name: '{{ wiki_db_container_name }}'
     detach: true
     restart_policy: '{{ docker_restart_policy }}'
-    volumes_from: '{{ wiki_db_dv_name }}'
-    image: '{{ wiki_db_image_name }}:{{ docker_image_tag }}'
+    image: '{{ wiki_db_image_name }}'
+    volumes:
+      - '{{ wiki_db_dv_name }}:/var/lib/mysql:z'
+    env:
+      MYSQL_DATABASE: '{{ wiki_database_name }}'
+      MYSQL_ROOT_PASSWORD: '{{ wiki_db_root_password }}'
+      MYSQL_USER: '{{ wiki_db_user }}'
+      MYSQL_PASSWORD: '{{ wiki_db_password }}'
+    state: started
+
+- name: wait for mysql initialization to complete
+  shell: >
+      printf 'SHOW GLOBAL STATUS\n' | docker exec -i 
+      '{{ wiki_db_container_name }}'
+      mysql --host=localhost
+      --user={{ wiki_db_user | quote }}
+      --password={{ wiki_db_password | quote }}
+      '{{ wiki_database_name }}'
+  register: db_status
+  retries: 20
+  delay: 3
+  until: "db_status.rc == 0"
+
+# *****************************************************************************
+# Start the data container running
 
 - name: start container (mediwiki)
   docker_container:
@@ -137,9 +139,9 @@
       - '{{ openldap_container_name }}:wiki_ldap'
       - '{{ wiki_db_container_name }}:wiki_db'
       - '{{ exim4_container_name }}:{{ container_addr_map.exim4.hostname }}'
-    volumes_from: 
-      - '{{ openssl_dv_name }}'
-      - '{{ wiki_dv_name }}'
+    volumes:
+      - '{{ openssl_dv_name }}:/etc/ssl/private:z'
+      - '{{ wiki_dv_name }}/var/www/html:z'
     env:
       WIKI_HOSTNAME: '{{ container_addr_map.wiki.hostname }}'
       WIKI_MAIL_USER: 'wiki'
diff --git a/roles/docker-mediawiki/templates/config.sh.j2 b/roles/docker-mediawiki/templates/config.sh.j2
index 28f6a40f..4e6831e3 100644
--- a/roles/docker-mediawiki/templates/config.sh.j2
+++ b/roles/docker-mediawiki/templates/config.sh.j2
@@ -10,5 +10,5 @@
   HOST_WIKI_BACKUP_DIR='{{ wiki_docker_backup_dir }}'
   STATIC_BACKUP_FILE='{{ wiki_static_backup_file }}'
   DATABASE_BACKUP_FILE='{{ wiki_database_backup_file }}'
-  WIKI_DB_DB_NAME='wikidb'
+  WIKI_DB_DB_NAME='{{ wiki_database_name }}'
 }
diff --git a/roles/docker-openldap/defaults/main.yml b/roles/docker-openldap/defaults/main.yml
index dd0484f2..1a00928d 100644
--- a/roles/docker-openldap/defaults/main.yml
+++ b/roles/docker-openldap/defaults/main.yml
@@ -34,6 +34,7 @@ openldap_image_name: '{{ docker_registry_username }}/ldap'
 openldap_container_name: '{{ docker_name_prefix }}ldap'
 
 # the name of the data-volume used by the container
+openldap_cv_name: '{{ docker_name_prefix }}openldap_CV'
 openldap_dv_name: '{{ docker_name_prefix }}openldap_DV'
 
 # restore directories to temporarly store data being restored into docker containers
diff --git a/roles/docker-openldap/tasks/main.yml b/roles/docker-openldap/tasks/main.yml
index 1ef879c3..b603dd88 100644
--- a/roles/docker-openldap/tasks/main.yml
+++ b/roles/docker-openldap/tasks/main.yml
@@ -76,11 +76,13 @@
 # *****************************************************************************
 # Create the data volumes
 
-- name: data-volume container
-  docker_datavolume:
-    image_name: '{{ openldap_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ openldap_dv_name }}'
+- name: config volume
+  docker_volume:
+    name: '{{ openldap_cv_name }}'
+
+- name: data volume
+  docker_volume:
+    name: '{{ openldap_dv_name }}'
 
 # *****************************************************************************
 # Populate the ldap data-volume with default data
@@ -92,7 +94,8 @@
     image: '{{ openldap_image_name }}:{{ docker_image_tag }}'
     command: ['init_data_volumes']
     volumes:
-      - '{{ openldap_dv_name }}'
+      - '{{ openldap_cv_name }}:/etc/ldap:z'
+      - '{{ openldap_dv_name }}:/var/lib/ldap:z'
   when: st_openldap_restore.stat.exists == False
 
 - name: initial populate
@@ -110,9 +113,10 @@
     restart_policy: '{{ docker_restart_policy }}'
     hostname: '{{ container_addr_map.ldap.hostname }}'
     ports: '{{ container_port_map.ldap.port_args }}'
-    volumes_from:
-      - '{{ openssl_dv_name }}'
-      - '{{ openldap_dv_name }}'
+    volumes:
+      - '{{ openssl_dv_name }}:/etc/ssl/private:z'
+      - '{{ openldap_cv_name }}:/etc/ldap:z'
+      - '{{ openldap_dv_name }}:/var/lib/ldap:z'
     env:
       LDAP_HOSTNAME: '{{ container_addr_map.ldap.hostname }}'
     image: '{{ openldap_image_name }}:{{ docker_image_tag }}'
@@ -124,8 +128,8 @@
     restart_policy: '{{ docker_restart_policy }}'
     hostname: '{{ container_addr_map.ldap.hostname }}'
     ports: '{{ container_port_map.phpldapadmin.port_args }}'
-    volumes_from:
-      - '{{ openssl_dv_name }}'
+    volumes:
+      - '{{ openssl_dv_name }}:/etc/ssl/private:z'
     env:
       PHPLDAPADMIN_HTTPS: 'false'
       PHPLDAPADMIN_LDAP_HOSTS: 'ldap'
diff --git a/roles/docker-openldap/tasks/restore.yml b/roles/docker-openldap/tasks/restore.yml
index 370f4797..d139794e 100644
--- a/roles/docker-openldap/tasks/restore.yml
+++ b/roles/docker-openldap/tasks/restore.yml
@@ -34,7 +34,7 @@
   command: 'docker stop {{ openldap_container_name }}'
 
 - name: load ldif files
-  command: 'docker run --rm --volumes-from {{ openldap_dv_name }} -v {{ host_openldap_docker_restore_dir }}:/tmp/import_export {{ openldap_image_name }}:{{ docker_image_tag }} apply_ldif -n {{ item.dbnum }} -l {{ item.ldif_file }}'
+  command: 'docker run --rm --volume {{ openldap_cv_name }}:/etc/ldap:z --volume {{ openldap_dv_name }}:/var/lib/ldap:z -v {{ host_openldap_docker_restore_dir }}:/tmp/import_export {{ openldap_image_name }}:{{ docker_image_tag }} apply_ldif -n {{ item.dbnum }} -l {{ item.ldif_file }}'
   with_items: '{{ openldap_backup_files }}'
 
 - name: restart container (ldap)
diff --git a/roles/docker-openssl/tasks/main.yml b/roles/docker-openssl/tasks/main.yml
index 5522f437..53d4b541 100644
--- a/roles/docker-openssl/tasks/main.yml
+++ b/roles/docker-openssl/tasks/main.yml
@@ -69,11 +69,9 @@
 # *****************************************************************************
 # Create the data volumes
 
-- name: data-volume container
-  docker_datavolume:
-    image_name: '{{ openssl_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ openssl_dv_name }}'
+- name: certificates volume
+  docker_volume:
+    name: '{{ openssl_dv_name }}'
 
 # *****************************************************************************
 # Populate the openssl data-volume with default data
diff --git a/roles/docker-openssl/templates/before_backup.j2 b/roles/docker-openssl/templates/before_backup.j2
index 160ff326..75fb3aca 100644
--- a/roles/docker-openssl/templates/before_backup.j2
+++ b/roles/docker-openssl/templates/before_backup.j2
@@ -5,7 +5,8 @@
 docker run --rm -v {{ docker_backup_dir }}:/tmp/import_export:z {{ repository_images.debian.image }}:{{ repository_images.debian.tag }} bash -c "rm -rf /tmp/import_export/openssl"
 mkdir -p -m u=rwx,g=rwx,o= {{ openssl_docker_backup_dir }}
 docker run --rm \
-    --volumes-from {{ openssl_dv_name }} \
+    --volume {{ openssl_dv_name }}:/etc/ssl/private:z \
     {{ openssl_image_name }}:{{ docker_image_tag }} backup > \
         {{ openssl_docker_backup_dir }}{{ openssl_backup_file }}
+
 {% endif %}
diff --git a/roles/docker-svn/tasks/main.yml b/roles/docker-svn/tasks/main.yml
index 5110de79..efa9b7e9 100644
--- a/roles/docker-svn/tasks/main.yml
+++ b/roles/docker-svn/tasks/main.yml
@@ -69,10 +69,8 @@
 # Create the data volumes
 
 - name: data-volume container
-  docker_datavolume:
-    image_name: '{{ svn_image_name }}'
-    image_tag: '{{ docker_image_tag }}'
-    data_volume_container_name: '{{ svn_dv_name }}'
+  docker_volume:
+    name: '{{ svn_dv_name }}'
 
 # *****************************************************************************
 # Start the data container running
@@ -86,9 +84,9 @@
     ports: '{{ container_port_map.svn.port_args }}'
     link:
       - '{{ openldap_container_name }}:ldap'
-    volumes_from:
-      - '{{ openssl_dv_name }}'
-      - '{{ svn_dv_name }}'
+    volumes:
+      - '{{ openssl_dv_name }}:/etc/ssl/private:z'
+      - '{{ svn_dv_name }}:/var/lib/svn:z'
     env:
       SVN_HOSTNAME: '{{ container_addr_map.svn.hostname }}'
       SVN_LDAP_PASSWORD: '{{ openldap_proxyagent_password }}'
diff --git a/roles/docker-svn/tasks/restore.yml b/roles/docker-svn/tasks/restore.yml
index f17b49bc..0a3fa25a 100644
--- a/roles/docker-svn/tasks/restore.yml
+++ b/roles/docker-svn/tasks/restore.yml
@@ -38,9 +38,8 @@
 - name: restore svn repositories
   docker_container:
     detach: False
-    volumes_from:
-      - '{{ svn_dv_name }}'
     volumes:
+      - '{{ svn_dv_name }}:/var/lib/svn:z'
       - '{{ svn_docker_restore_dir }}:/tmp/import_export:z'
     image: '{{ svn_image_name }}:{{ docker_image_tag }}'
     command: ['import', '{{ svn_repos }}']
